Runtime tests fail to AOT compile on Apple mobile platforms

## Build Information
Build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424
Build error leg or test failing: https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424&view=logs&j=da83f4c3-af55-5024-34f6-3ac8d11ac15f
<!-- Error message template  -->
## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssueJsonStepByStep.md).

<!-- Use ErrorMessage for String.Contains matches. Use ErrorPattern for regex matches (single line/no backtracking). Set BuildRetry to `true` to retry builds with this error. Set ExcludeConsoleLog to `true` to skip helix logs analysis. -->

```json
{
  "ErrorMessage": "duplicate symbol '_dn_simdhash_ptrpair_ptr_meta'",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

```
...
[/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: duplicate symbol '_dn_simdhash_ptrpair_ptr_vtable' in: [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmono-component-diagnostics_tracing-static.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmonosgen-2.0.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: duplicate symbol '_dn_simdhash_ptrpair_ptr_try_add' in: [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmono-component-diagnostics_tracing-static.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmonosgen-2.0.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: duplicate symbol '_dn_simdhash_ptrpair_ptr_meta' in: [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmono-component-diagnostics_tracing-static.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018:     /tmp/helix/working/95610881/w/A2B908C5/e/publish/libmonosgen-2.0.a(dn-simdhash-ptrpair-ptr.c.o) [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
/tmp/helix/working/95610881/p/build/apple/AppleBuild.targets(296,5): error MSB4018: ld: 156 duplicate symbols for architecture arm64 [/private/tmp/helix/working/95610881/w/A2B908C5/e/publish/ProxyProjectForAOTOnHelix.proj]
```

The regression is likely introduced in https://github.com/dotnet/runtime/pull/103184.

/cc: @jkoritzinsky 
<!-- Known issue validation start -->
 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424
**Error message validated:** `[duplicate symbol '_dn_simdhash_ptrpair_ptr_meta'`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 6/25/2024 6:36:10 PM UTC
<!-- Known issue validation end -->
<!--Known issue error report start -->

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[722476](https://dev.azure.com/dnceng-public/public/_build/results?buildId=722476)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=722476&view=ms.vss-test-web.build-test-results-tab&runId=18129490&resultId=100013)||
|[721722](https://dev.azure.com/dnceng-public/public/_build/results?buildId=721722)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=721722&view=ms.vss-test-web.build-test-results-tab&runId=18107258&resultId=100012)||
|[720898](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720898)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720898&view=ms.vss-test-web.build-test-results-tab&runId=18081448&resultId=100013)||
|[720816](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720816)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=720816&view=ms.vss-test-web.build-test-results-tab&runId=18078286&resultId=100013)||
|[719424](https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424)|dotnet/runtime|[baseservices_TieredCompilation.WorkItemExecution](https://dev.azure.com/dnceng-public/public/_build/results?buildId=719424&view=ms.vss-test-web.build-test-results-tab&runId=18033732&resultId=100013)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|5|5|
<!--Known issue error report end -->

--- a/src/mono/mono/component/CMakeLists.txt
+++ b/src/mono/mono/component/CMakeLists.txt
@@ -228,7 +228,11 @@ if (ENABLE_PERFTRACING AND "${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}" IN_LIST 
     # Link the diagnostics_tracing component to the EventPipe and DiagnosticServer libraries.
     if (TARGET ${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-objects)
         target_sources(${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-objects PRIVATE $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)
-        list(APPEND mono-components-objects $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)
+        # If we are going to link the diagnostics component via mono-components-objects, link in the supporting object libraries as well.
+        string(REPLACE
+          "$<TARGET_OBJECTS:${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-objects>"
+          "$<TARGET_OBJECTS:${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-objects>;$<TARGET_OBJECTS:eventpipe-mono-objects>;$<TARGET_OBJECTS:dn-containers>"
+          mono-components-objects "${mono-components-objects}")
     endif()
     if (TARGET mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-static)
         # We don't add dn-containers here as any scenario that uses this target will also use the monosgen-static target,

--- a/src/mono/mono/component/CMakeLists.txt
+++ b/src/mono/mono/component/CMakeLists.txt
@@ -231,7 +231,9 @@ if (ENABLE_PERFTRACING AND "${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}" IN_LIST 
         list(APPEND mono-components-objects $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)
     endif()
     if (TARGET mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-static)
-        target_sources(mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-static PRIVATE $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)
+        # We don't add dn-containers here as any scenario that uses this target will also use the monosgen-static target,
+        # which will also have dn-containers linked into it.
+        target_sources(mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME}-static PRIVATE $<TARGET_OBJECTS:eventpipe-mono-objects>)
     endif()
     if (TARGET mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME})
         target_sources(mono-component-${MONO_DIAGNOSTICS_TRACING_COMPONENT_NAME} PRIVATE $<TARGET_OBJECTS:eventpipe-mono-objects> $<TARGET_OBJECTS:dn-containers>)

